diffsmhm/loader.py

```diff
@@ -73,6 +79,366 @@ def wrap_to_local_volume_inplace(pos, cen, lbox):
+        If True, mmhid and mmh_x/y/z are added to halo_file
+
+    Returns
+    -------
+    mmhid : array like
+        most massive host of a each halo; may be an "indirect" host.
+    mmh_x : array like
+        x position of each halo's most massive host
+    mmh_y : array_like
+        y position of each halo's most massive host
+    mmh_z : array_like
+        z position of each halo's most massive host
+    mmh_dist : array_like
+        distance between each halo and it's mmh
+    """
+    # note these are different from load_and_chop
+    important_keys = ["halo_id", "upid", "pid", "mpeak", "x", "y", "z",
+                      "host_x", "host_y", "host_z", "host_dist", "rvir"]
+
+    # load hdf5 data
+    halos = OrderedDict()
+    with h5py.File(halo_file, "r") as hdf:
+        _host_mpeak_mask = np.log10(hdf["host_mpeak"][...]) >= host_mpeak_cut
+        for key in hdf.keys():
+            # only keep the columns that we want
+            if key not in important_keys:
+                continue
+
+            # integer types
+            if key in ("halo_id", "upid", "pid"):
+                dt = "i8"
+            else:
+                dt = "f4"
+            halos[key] = hdf[key][...][_host_mpeak_mask].astype(dt)
+
+    # 1) Point independent hosts that share a sub with a larger host to that larger host
+    # only consider subhalos with pid != upid
+    pid_upid_diff = halos["pid"][halos["pid"] != halos["upid"]]
+
+    # get list of indices of "pid hosts"
+    host_indices = np.where(halos["upid"] == -1)[0]
+    pid_indices = np.where(np.isin(halos["halo_id"], pid_upid_diff))[0]
+    pid_hosts_indices = np.intersect1d(host_indices, pid_indices)
+
+    # now each mpi rank determines it's chunk of the list of "pid hosts"
+    avg, rem = divmod(len(pid_hosts_indices), N_RANKS)
+    rank_count = [avg + 1 if p < rem else avg for p in range(N_RANKS)]
+    displ = [sum(rank_count[:p]) for p in range(N_RANKS)]
+
+    rank_count = rank_count[RANK]
+    displ = displ[RANK]
+
+    pid_hosts_indices_rank = pid_hosts_indices[displ:displ+rank_count]
+
+    # get id's and upid's for this rank
+    pid_hosts_rank = np.copy(halos["halo_id"][pid_hosts_indices_rank])
+    upid_corr_rank = np.copy(halos["upid"][pid_hosts_indices_rank])
+
+    # determine and assign new upids for "pid hosts"
+    for i, host in enumerate(pid_hosts_rank):
+        mask_this_host = halos["halo_id"] == host
+
+        # get all subs that have this host as their pid
+        subs_this_pid = halos["halo_id"][halos["pid"] == host]
+
+        # get all upids of host's subs
+        upids_poss = np.unique(halos["upid"][np.isin(halos["halo_id"], subs_this_pid)])
+        # remove any subs from the upid list (otherwise can get looping)
+        upids_poss = np.setdiff1d(upids_poss, subs_this_pid)
+        # remove host from the upid list (otherwise can get looping)
+        upids_poss = np.setdiff1d(upids_poss, host)
+        if len(upids_poss) == 0:
+            continue
```

Comment:
What kind of halo configuration triggers this branch?

Reply:
If there is a sub halo that has the same host as its pid and upid, then naively we could create a loop. In removing the host from the list of possible upids, it's possible there are no other candidate upids, hence the length check. 

It's entirely possible there's a more elegant way to handle that situation; I'll revisit this logic